homeassistant/components/home_plus_control/api.py

```diff
@@ -0,0 +1,352 @@
+                                                       token refresh.
+        logger (Logger): Logger of the object.
+        switches (dict): Dictionary of HomePlusControl switches indexed by their unique ID.
+        switches_to_remove (dict): Dictionary of the HomePlusControl switches that are to be
+                                   removed from HomeAssistant, indexed by their unique ID.
+    """
+
+    def __init__(
+        self,
+        hass: core.HomeAssistant,
+        config_entry: config_entries.ConfigEntry,
+        implementation: config_entry_oauth2_flow.AbstractOAuth2Implementation,
+    ):
+        """Initialize the HomePlusControlAsyncApi object.
+
+        Initialize the authenticated API for the Legrand Home+ Control component.
+
+        Args:.
+            hass (HomeAssistant): HomeAssistant core object.
+            config_entry (ConfigEntry): ConfigEntry object that configures this API.
+            implementation (AbstractOAuth2Implementation): OAuth2 implementation that handles AA
+                                                           and token refresh.
+        """
+        self.hass = hass
+        self.config_entry = config_entry
+        self.implementation = implementation
+        self._domain = DOMAIN
+        self._plants = {}
+        self._switches = {}
+        self._switches_to_remove = {}
+
+        self._last_check = {
+            CONF_PLANT_UPDATE_INTERVAL: time.monotonic(),
+            CONF_PLANT_TOPOLOGY_UPDATE_INTERVAL: -1,
+            CONF_MODULE_STATUS_UPDATE_INTERVAL: -1,
+        }
+
+        self._refresh_intervals = {}
+        self.update_refresh_intervals()
+
+        # Create the API authenticated client - external library
+        super().__init__(
+            client_id=self.config_entry.data[CONF_CLIENT_ID],
+            client_secret=self.config_entry.data[CONF_CLIENT_SECRET],
+            subscription_key=self.config_entry.data[CONF_SUBSCRIPTION_KEY],
+            token=self.config_entry.data["token"],
+            oauth_client=aiohttp_client.async_get_clientsession(self.hass),
+        )
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Logger of authentication API."""
+        return logging.getLogger(__name__)
+
+    @property
+    def switches(self):
+        """Return dictionary of switch modules of this platform."""
+        return self._switches
+
+    @switches.setter
+    def switches(self, switches):
+        """Set the internal switch attribute."""
+        self._switches = switches
+
+    @property
+    def switches_to_remove(self):
+        """Return dictionary of switch modules of this platform that should be removed from HA."""
+        return self._switches_to_remove
+
+    @switches_to_remove.setter
+    def switches_to_remove(self, switches):
+        """Set the internal switches_to_remove attribute."""
+        self._switches_to_remove = switches
+
+    async def async_ensure_token_valid(self) -> None:
```

Comment:
This method is implemented by the Home Assistant oauth session. The library should instead implement an abstract method `async_get_access_token`.

Then here we should define:

```py
async def async_get_access_token(self) -> str:
    """Return a valid access token."""
    if not self._oauth_session.valid_token:
        await self._oauth_session.async_ensure_token_valid()

    return self._oauth_session.token["access_token"]
```

https://developers.home-assistant.io/docs/api_lib_auth#async-example-1